File: IntroToF#/src/FSharp/TypeInference.fs
module TypeInference

    open System.Net

    // Where have all the types gone? Done need 'em!

    let getFormattedWebPage url formatter = 
        let webClient = new WebClient()
        let html = webClient.DownloadString(url : string)
        (url, formatter html)

    let url = "http://www.google.com"

    let googleHomePageHtml = getFormattedWebPage url (fun x -> x.ToUpper())



    // Automatic Generalization

    let greaterThan a b = a > b
    // val greaterThan: 'a -> 'a -> bool 

    let add x y = x + y
    // val add: int -> int -> int

File: IntroToF#/src/CSharp/TypeInference.cs
using System;
using System.Net;

namespace CSharp
{
    public class TypeInference
    {
        string GetFormatedWebPage(string url, Func<string, string> formatter)
        {
            WebClient webClient = new WebClient();
            string html = webClient.DownloadString(url);
            return formatter(html);
        }

        void Sample2()
        {
            string url = "http://www.googl.com";

            string googleHomePage = GetFormatedWebPage(url, delegate(string x) { return x.ToUpper(); });

            Tuple<string, string> webPage = new Tuple<string, string>(url, googleHomePage);
        }

        // -----------------------------------------------------------------------------------------------------

        string GetFormatedWebPage2(string url, Func<string, string> formatter)
        {
            var webClient = new WebClient();
            var html = webClient.DownloadString(url);
            return formatter(html);
        }

        void Sample3()
        {
            var url = "http://www.googl.com";

            var googleHomePage = GetFormatedWebPage(url, x => x.ToUpper());

            var webPage = new Tuple<string,string>(url, googleHomePage);
        }
    }
}
